scipy/stats/_hypotests.py

```diff
@@ -1262,30 +1262,46 @@ def _get_binomial_log_p_value_with_nuisance_param(
     return -log_pvalue
 
 
-def _pval_cvm_2samp_exact(s, nx, ny):
+def _pval_cvm_2samp_exact(s, m, n):
     """
     Compute the exact p-value of the Cramer-von Mises two-sample test
-    for a given value s (float) of the test statistic by enumerating
-    all possible combinations. nx and ny are the sizes of the samples.
+    for a given value s of the test statistic.
+    m and n are the sizes of the samples.
+
+    [1] Y. Xiao, A. Gordon, and A. Yakovlev, "A C++ Program for
+        the Cramér-Von Mises Two-Sample Test", J. Stat. Soft.,
+        vol. 17, no. 8, pp. 1-15, Dec. 2006.
+    [2] T. W. Anderson "On the Distribution of the Two-Sample Cramer-von Mises
+        Criterion," The Annals of Mathematical Statistics, Ann. Math. Statist.
+        33(3), 1148-1159, (September, 1962)
     """
-    rangex = np.arange(nx)
-    rangey = np.arange(ny)
 
-    us = []
-
-    # x and y are all possible partitions of ranks from 0 to nx + ny - 1
-    # into two sets of length nx and ny
-    # Here, ranks are from 0 to nx + ny - 1 instead of 1 to nx + ny, but
-    # this does not change the value of the statistic.
-    for x, y in _all_partitions(nx, ny):
-        # compute the statistic
-        u = nx * np.sum((x - rangex)**2)
-        u += ny * np.sum((y - rangey)**2)
-        us.append(u)
-
-    # compute the values of u and the frequencies
-    u, cnt = np.unique(us, return_counts=True)
-    return np.sum(cnt[u >= s]) / np.sum(cnt)
+    # [1, p. 3]
+    lcm = np.lcm(m, n)
+    # [1, p. 4], below eq. 3
+    a = lcm // m
+    b = lcm // n
+    # eq. 9 in [2] and eq. 2 in [1]
+    mn = m * n
+    zeta = lcm ** 2 * (m + n) * (6 * s - mn * (4 * mn - 1)) // (6 * mn ** 2)
+    # the frequency table of $g_{u, v}^+$ defined in [1, p. 6]
+    gs = [np.array([[0], [1]])] + [np.empty((2, 0), int) for _ in range(m)]
+    for u in range(n + 1):
+        next_gs = []
+        tmp = np.empty((2, 0), int)
+        for v, g in enumerate(gs):
+            # calculate g recursively with eq. 11 in [1]
+            vi, i0, i1 = np.intersect1d(tmp[0], g[0], return_indices=True)
+            tmp = np.concatenate([
+                np.stack([vi, tmp[1, i0] + g[1, i1]]),
+                np.delete(tmp, i0, 1),
+                np.delete(g, i1, 1)
```

Comment:
My guess is that reshaping the array the inner loop will slow things down quite a lot. For instance, every time you do `delete` it will have to copy the whole table to a new location in memory (with one row removed). Can you pre-allocate the right amount of space and move only the things that _must_ be moved? Or would it make sense to use some other data structure that is more flexible?

Reply:
> would it make sense to use some other data structure that is more flexible (e.g. dict)?

I once thought dict (or `collections.Counter`) was suitable. But then we need to add a constant value to every key (not value) of a dict. This means every hash must be re-calculated, so I think this operation is slow. With NumPy arrray, we can write simply as `tmp[0] += (a * v - b * u) ** 2`.

> Can you pre-allocate the right amount of space and move only the things that must be moved?

I did not know that it was possible. I will try it.

Reply:
Well it depends on whether you can tell ahead of time how many elements the arrays will need. Or you could guess how many elements they'll need and double them whenever you need more. 
Another thing that might be faster is if `i0` and `i1` could be logical arrays. 
But this is not as slow (compared to what it could be) as I originally thought. We can leave improvements up to a follow-up PR. For now, let's just get the algorithm settled.